storage/gcp/gcp.go

```diff
@@ -217,7 +237,102 @@ func (s *Storage) setEntryBundle(ctx context.Context, bundleIndex uint64, logSiz
 
 // integrate incorporates the provided entries into the log starting at fromSeq.
 func (s *Storage) integrate(ctx context.Context, fromSeq uint64, entries [][]byte) error {
-	return nil
+	tb := storage.NewTreeBuilder(func(ctx context.Context, tileIDs []storage.TileID, treeSize uint64) ([]*api.HashTile, error) {
+		n, err := s.getTiles(ctx, tileIDs, treeSize)
+		if err != nil {
+			return nil, fmt.Errorf("getTiles: %w", err)
+		}
+		return n, nil
+	})
+
+	errG := errgroup.Group{}
+
+	errG.Go(func() error {
+		if err := s.updateEntryBundles(ctx, fromSeq, entries); err != nil {
+			return fmt.Errorf("updateEntryBundles: %v", err)
+		}
+		return nil
+	})
+
+	errG.Go(func() error {
+		newSize, newRoot, tiles, err := tb.Integrate(ctx, fromSeq, entries)
+		if err != nil {
+			return fmt.Errorf("Integrate: %v", err)
+		}
+		for k, v := range tiles {
+			func(ctx context.Context, k storage.TileID, v *api.HashTile) {
+				errG.Go(func() error {
+					return s.setTile(ctx, uint64(k.Level), k.Index, newSize, v)
+				})
+			}(ctx, k, v)
+		}
+		errG.Go(func() error {
+			//TODO: write out checkpoint
+			klog.Infof("New CP: %d, %x", newSize, newRoot)
+			return nil
+		})
+
+		return nil
+	})
+
+	return errG.Wait()
+}
+
+// updateEntryBundles adds the entries being integrated into the entry bundles.
+//
+// The right-most bundle will be grown, if it's partial, and/or new bundles will be created as required.
+func (s *Storage) updateEntryBundles(ctx context.Context, fromSeq uint64, entries [][]byte) error {
+	numAdded := uint64(0)
+	bundleIndex, entriesInBundle := fromSeq/entryBundleSize, fromSeq%entryBundleSize
+	bundle := &api.EntryBundle{}
+	if entriesInBundle > 0 {
+		// If the latest bundle is partial, we need to read the data it contains in for our newer, larger, bundle.
+		part, err := s.getEntryBundle(ctx, uint64(bundleIndex), uint64(entriesInBundle))
+		if err != nil {
+			return err
+		}
+		bundle = part
+	}
+
+	seqErr := errgroup.Group{}
+
+	// goSetEntryBundle is a function which uses seqErr to spin off a go-routine to write out an entry bundle.
+	// It's used in the for loop below.
+	goSetEntryBundle := func(ctx context.Context, bundleIndex uint64, fromSeq uint64, bundle api.EntryBundle) {
+		seqErr.Go(func() error {
+			if err := s.setEntryBundle(ctx, bundleIndex, fromSeq, &bundle); err != nil {
+				if !errors.Is(os.ErrExist, err) {
+					return err
+				}
```

Comment:
It could be worth logging something here when a duplicate write has happened. What I'd actually prefer would be prometheus counters for successful write, error write, duplicate write, etc. But that can come later?

Reply:
FWIW the GCP impl already logs a (verbose level) message when it encounters an idempotent dupe write, but agree that metrics are the right thing.